# Descriptive Statistics — Measures of Central Tendency and Variability

**Objective:**
1. Provide summary statistics (mean, median, min, max, standard deviation) for numeric variables grouped by a categorical variable. Create a list with a numeric value for each response of the categorical variable.
2. Display basic statistical details (percentile, mean, standard deviation) for each species in the Iris dataset (`Iris-setosa`, `Iris-versicolor`, `Iris-virginica`).

**Dataset:** Iris (`iris.csv`) — 150 rows, 5 columns (4 numeric, 1 categorical).

### Key Concepts

**Measures of Central Tendency** — describe the *center* of a distribution:
- **Mean** — arithmetic average; sensitive to outliers.
- **Median** — middle value when sorted; robust to outliers.
- **Mode** — most frequent value.

**Measures of Variability** — describe the *spread* of a distribution:
- **Minimum / Maximum** — smallest / largest value.
- **Range** — max − min.
- **Standard Deviation (std)** — average distance of each value from the mean.
- **Variance** — square of the standard deviation.
- **Percentile** — value below which a given % of observations fall (25th = Q1, 50th = median, 75th = Q3).

Descriptive Statistics (Iris Dataset)
Big picture: Notebook 1 and 2 were about cleaning the data. This one is about summarizing it — computing averages, spreads, and percentiles, particularly broken down per category.
What we did, step by step:

Brought in our toolbox — Just pandas and numpy this time. No models, no plots — pure summary statistics.
Loaded the Iris dataset — 150 flowers, four physical measurements (sepal length/width, petal length/width), and the species (Setosa, Versicolor, Virginica) — three categories with 50 flowers each.
Confirmed the structure — Verified the four measurements are numeric, the species column is categorical, and listed the three unique species names.

Part 1 — Summary Stats Grouped by Category

Grouped by species and computed each statistic separately:

Mean — the average measurement within each species.
Median — the middle measurement when sorted; useful when outliers exist.
Minimum and maximum — the smallest and largest values per species.
Standard deviation — how spread-out the measurements are within each species. A small number means flowers in that species are very similar to each other; a large number means they vary a lot.


Combined all stats in a single table — Same numbers as above, but stacked into one tidy table for easy comparison across species.
Built the required list — The assignment specifically asked for "a list with one numeric value per categorical response". We computed the mean sepal length for each species, giving exactly three numbers — one per species. Also showed it as a dictionary mapping each species name to its number for clarity.

Part 2 — Per-Species Statistical Details

Filtered the data three times, once for each species, and on each subset asked Python for a full statistical summary in one shot. This single command returns:

Count (how many flowers in this species)
Mean
Standard deviation
Minimum
25th percentile (Q1) — value below which 25% of measurements fall
50th percentile (Q2) — same as median
75th percentile (Q3) — value below which 75% of measurements fall
Maximum


Showed the one-line shortcut — Same result for all three species at once by combining grouping + summarizing in a single instruction.
Concluded with what we observed: Setosa has the smallest petals, Virginica the largest, Versicolor sits between them. Standard deviations are small within each species — measurements cluster tightly — which is precisely why Iris is the textbook "easy classification" dataset.


## Step 1: Import Libraries

In [2]:
# pandas — DataFrame operations: read_csv(), groupby(), describe(), agg().
import pandas as pd

# numpy — numerical operations and array support (used internally by pandas).
import numpy as np

# Display all columns in output
pd.set_option('display.max_columns', None)

## Step 2: Load the Dataset

In [ ]:
# read_csv() loads a CSV file into a pandas DataFrame.
df = pd.read_csv('Descriptive_Statistics_ I I I.csv')
# df1 = pd.read_csv('Descriptive_Statistics_ I I I.csv')  # IGNORE
# shape -> (rows, columns)
print("Shape:", df.shape)

# head() shows the first 5 rows.
df.head()

Shape: (150, 5)


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


In [ ]:
# dtypes shows the data type of each column:
#   float64 -> numeric (continuous)
#   object  -> string (categorical 'species')
print(df.dtypes)
# print(df.dtypes)
# Verify the categorical variable's unique categories.
print("\nSpecies categories:", df['species'].unique())

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species             str
dtype: object

Species categories: <StringArray>
['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
Length: 3, dtype: str


## Part 1: Summary Statistics Grouped by a Categorical Variable

We treat **`species`** as the categorical (qualitative) variable and the four flower measurements as the numeric (quantitative) variables. We then compute summary statistics for each numeric variable *within each species*.

This is exactly analogous to the textbook example "income grouped by age groups" — we just have "petal/sepal measurements grouped by species".

### 1.1 Group by Species and Compute Summary Statistics

In [5]:
# groupby('species') splits the DataFrame into 3 groups (one per species).
# Calling an aggregation function then computes that statistic for each group.

# MEAN — arithmetic average of each numeric column, per species.
print("MEAN of each numeric variable, grouped by species:\n")
print(df.groupby('species').mean(numeric_only=True))

MEAN of each numeric variable, grouped by species:

                 sepal_length  sepal_width  petal_length  petal_width
species                                                              
Iris-setosa             5.006        3.428         1.462        0.246
Iris-versicolor         5.936        2.770         4.260        1.326
Iris-virginica          6.588        2.974         5.552        2.026


In [5]:
# MEDIAN — middle value (50th percentile) of each numeric column, per species.
# More robust than mean when the data has outliers.
print("MEDIAN of each numeric variable, grouped by species:\n")
print(df.groupby('species').median(numeric_only=True))

MEDIAN of each numeric variable, grouped by species:

                 sepal_length  sepal_width  petal_length  petal_width
species                                                              
Iris-setosa               5.0          3.4          1.50          0.2
Iris-versicolor           5.9          2.8          4.35          1.3
Iris-virginica            6.5          3.0          5.55          2.0


In [6]:
# MIN — smallest value in each group.
print("MINIMUM of each numeric variable, grouped by species:\n")
print(df.groupby('species').min(numeric_only=True))

MINIMUM of each numeric variable, grouped by species:

                 sepal_length  sepal_width  petal_length  petal_width
species                                                              
Iris-setosa               4.3          2.3           1.0          0.1
Iris-versicolor           4.9          2.0           3.0          1.0
Iris-virginica            4.9          2.2           4.5          1.4


In [7]:
# MAX — largest value in each group.
print("MAXIMUM of each numeric variable, grouped by species:\n")
print(df.groupby('species').max(numeric_only=True))

MAXIMUM of each numeric variable, grouped by species:

                 sepal_length  sepal_width  petal_length  petal_width
species                                                              
Iris-setosa               5.8          4.4           1.9          0.6
Iris-versicolor           7.0          3.4           5.1          1.8
Iris-virginica            7.9          3.8           6.9          2.5


In [8]:
# STANDARD DEVIATION — measures spread of values around the mean within each group.
# A small std means values are tightly clustered; a large std means they are spread out.
print("STANDARD DEVIATION of each numeric variable, grouped by species:\n")
print(df.groupby('species').std(numeric_only=True))

STANDARD DEVIATION of each numeric variable, grouped by species:

                 sepal_length  sepal_width  petal_length  petal_width
species                                                              
Iris-setosa          0.352490     0.379064      0.173664     0.105386
Iris-versicolor      0.516171     0.313798      0.469911     0.197753
Iris-virginica       0.635880     0.322497      0.551895     0.274650


### 1.2 All Statistics in One Table — agg()

In [10]:
# agg() applies multiple aggregation functions at once and presents them in
# one tidy table. Far more compact than calling mean(), median(), etc. separately.
summary = df.groupby('species').agg(['mean', 'median', 'min', 'max', 'std'])

print("Combined summary statistics grouped by species:")
summary

Combined summary statistics grouped by species:


sepal_length                            sepal_width         \
                        mean median  min  max       std        mean median   
species                                                                      
Iris-setosa            5.006    5.0  4.3  5.8  0.352490       3.428    3.4   
Iris-versicolor        5.936    5.9  4.9  7.0  0.516171       2.770    2.8   
Iris-virginica         6.588    6.5  4.9  7.9  0.635880       2.974    3.0   

                                    petal_length                             \
                 min  max       std         mean median  min  max       std   
species                                                                       
Iris-setosa      2.3  4.4  0.379064        1.462   1.50  1.0  1.9  0.173664   
Iris-versicolor  2.0  3.4  0.313798        4.260   4.35  3.0  5.1  0.469911   
Iris-virginica   2.2  3.8  0.322497        5.552   5.55  4.5  6.9  0.551895   

                petal_width                             
                       mean median  min  max       std  
species                                                 
Iris-setosa           0.246    0.2  0.1  0.6  0.105386  
Iris-versicolor       1.326    1.3  1.0  1.8  0.197753  
Iris-virginica        2.026    2.0  1.4  2.5  0.274650

### 1.3 Create a List of Numeric Values for Each Categorical Response

The assignment asks: *Create a list that contains a numeric value for each response to the categorical variable.*

We build a list where each element is the **mean `sepal_length`** of one species. So the list has exactly 3 numbers — one numeric value per category (species).

In [10]:
# tolist() converts a pandas Series into a regular Python list.
# Here we get the mean sepal_length for each of the 3 species as a list.
mean_sepal_length_per_species = df.groupby('species')['sepal_length'].mean().tolist()

print("Categories (species):", df['species'].unique().tolist())
print("Mean sepal_length per species:", mean_sepal_length_per_species)

Categories (species): ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
Mean sepal_length per species: [5.006, 5.936, 6.587999999999999]


In [11]:
# We can also build a dictionary {category -> numeric value} for clearer mapping.
# to_dict() turns a Series into a Python dict.
species_means = df.groupby('species')['sepal_length'].mean().to_dict()
print("Mean sepal_length per species (as dict):")
for species, mean_val in species_means.items():
    print(f"  {species}: {round(mean_val, 3)}")

Mean sepal_length per species (as dict):
  Iris-setosa: 5.006
  Iris-versicolor: 5.936
  Iris-virginica: 6.588


## Part 2: Statistical Details for Each Iris Species

The assignment asks for percentile, mean, and standard deviation for each species. The fastest way is `describe()`, which gives all of these (count, mean, std, min, 25%, 50%, 75%, max) in one call.

### 2.1 Iris-setosa

In [12]:
# Filter the DataFrame to only Iris-setosa rows.
# df['species'] == 'Iris-setosa' returns a boolean mask;
# df[mask] selects the rows where the mask is True.
setosa = df[df['species'] == 'Iris-setosa']

# describe() returns: count, mean, std, min, 25%, 50% (median), 75%, max.
# That single call covers ALL the statistics the assignment asks for.
print("Statistical details for Iris-setosa:")
setosa.describe()

Statistical details for Iris-setosa:


,sepal_length,sepal_width,petal_length,petal_width
count,50.00000,50.000000,50.000000,50.000000
mean,5.00600,3.428000,1.462000,0.246000
std,0.35249,0.379064,0.173664,0.105386
min,4.30000,2.300000,1.000000,0.100000
25%,4.80000,3.200000,1.400000,0.200000
50%,5.00000,3.400000,1.500000,0.200000
75%,5.20000,3.675000,1.575000,0.300000
max,5.80000,4.400000,1.900000,0.600000


### 2.2 Iris-versicolor

In [13]:
versicolor = df[df['species'] == 'Iris-versicolor']
print("Statistical details for Iris-versicolor:")
versicolor.describe()

Statistical details for Iris-versicolor:


,sepal_length,sepal_width,petal_length,petal_width
count,50.000000,50.000000,50.000000,50.000000
mean,5.936000,2.770000,4.260000,1.326000
std,0.516171,0.313798,0.469911,0.197753
min,4.900000,2.000000,3.000000,1.000000
25%,5.600000,2.525000,4.000000,1.200000
50%,5.900000,2.800000,4.350000,1.300000
75%,6.300000,3.000000,4.600000,1.500000
max,7.000000,3.400000,5.100000,1.800000


### 2.3 Iris-virginica

In [14]:
virginica = df[df['species'] == 'Iris-virginica']
print("Statistical details for Iris-virginica:")
virginica.describe()

Statistical details for Iris-virginica:


,sepal_length,sepal_width,petal_length,petal_width
count,50.00000,50.000000,50.000000,50.00000
mean,6.58800,2.974000,5.552000,2.02600
std,0.63588,0.322497,0.551895,0.27465
min,4.90000,2.200000,4.500000,1.40000
25%,6.22500,2.800000,5.100000,1.80000
50%,6.50000,3.000000,5.550000,2.00000
75%,6.90000,3.175000,5.875000,2.30000
max,7.90000,3.800000,6.900000,2.50000


### 2.4 One-line Alternative — All Three at Once

In [15]:
# Same result for ALL three species at once via groupby + describe.
df.groupby('species').describe()

sepal_length                                              \
                       count   mean       std  min    25%  50%  75%  max   
species                                                                    
Iris-setosa             50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8   
Iris-versicolor         50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0   
Iris-virginica          50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9   

                sepal_width                                                \
                      count   mean       std  min    25%  50%    75%  max   
species                                                                     
Iris-setosa            50.0  3.428  0.379064  2.3  3.200  3.4  3.675  4.4   
Iris-versicolor        50.0  2.770  0.313798  2.0  2.525  2.8  3.000  3.4   
Iris-virginica         50.0  2.974  0.322497  2.2  2.800  3.0  3.175  3.8   

                petal_length                                               \
                       count   mean       std  min  25%   50%    75%  max   
species                                                                     
Iris-setosa             50.0  1.462  0.173664  1.0  1.4  1.50  1.575  1.9   
Iris-versicolor         50.0  4.260  0.469911  3.0  4.0  4.35  4.600  5.1   
Iris-virginica          50.0  5.552  0.551895  4.5  5.1  5.55  5.875  6.9   

                petal_width                                            
                      count   mean       std  min  25%  50%  75%  max  
species                                                                
Iris-setosa            50.0  0.246  0.105386  0.1  0.2  0.2  0.3  0.6  
Iris-versicolor        50.0  1.326  0.197753  1.0  1.2  1.3  1.5  1.8  
Iris-virginica         50.0  2.026  0.274650  1.4  1.8  2.0  2.3  2.5

## Conclusion

**Part 1** — used `groupby('species')` with `.mean()`, `.median()`, `.min()`, `.max()`, `.std()` (and `.agg()` for the combined table) to summarize numeric variables per category. Built a list of mean `sepal_length` values, one per species.

**Part 2** — used boolean filtering (`df[df['species'] == ...]`) and `.describe()` to get count, mean, std, min, quartiles (25/50/75%), and max for each Iris species individually.

### Key Observations from the Iris Dataset
- `Iris-setosa` has the **smallest** petals (mean petal_length ≈ 1.46 cm) — easy to separate from the other two.
- `Iris-virginica` has the **largest** flowers in every dimension.
- `Iris-versicolor` sits between the other two species.
- Standard deviations are small within each species, meaning each species' measurements cluster tightly — which is why Iris is famously easy to classify.